tests/testthat/test-tree.R

```diff
@@ -0,0 +1,243 @@
+# Builds a regex that tests for the _lack_ of a substring
+doesNotExistRegex <- function(substring){
+  # Uses a negative look-behind to check for the absence of the substring
+  # pattern.
+
+  # The (?...) block controls how the perl regex parser works
+  # s -> Makes a dot matches all characters
+  # m -> Multiline mode
+  paste0("(?sm)^(?!.*", substring, ").*$")
+}
+
+
+# These tests are disabled because they require snapshot testing which is not available until testthat 3
+# test_that("Pure nested lists work", {
+#   plain_list <- list(
+#     list(id = "a",
+#          val = 2),
+#     list(id = "b",
+#          val = 1,
+#          children = list(
+#            list(id = "b1",
+#                 val = 2.5),
+#            list(id = "b2",
+#                 val = 8,
+#                 children = list(
+#                   list(id = "b21",
+#                        val = 4)
+#                 )))),
+#     list(id = "c",
+#          val = 8,
+#          children = list(
+#            list(),
+#            list(id = "c1"),
+#            list(id = "c2",
+#                 val = 1))))
+#
+#   expect_snapshot(tree(plain_list))
+#
+#   # Also with array-likes without indices
+#   expect_snapshot(tree(plain_list, index_arraylike = FALSE))
+# })
+#
+#
+# test_that("Atomic arrays have sensible defaults with truncation added for longer than 10-elements",{
+#   list_w_vectors <- list(
+#     name = "vectored list",
+#     num_vec = 1:10,
+#     char_vec = letters
+#   )
+#   expect_snapshot(
+#     tree(list_w_vectors)
+#   )
+# })
+#
+# test_that("Works with HTML tag structures", {
+#   # sliderInput is a pretty complex structure all in one line
+#   expect_snapshot(
+#     tree(shiny::sliderInput("test", "Input Label", 0,1,0.5))
+#   )
+# })
+
+test_that("Max depth can be enforced", {
```

Comment:
I think pretty much all of these should be converted to snapshot tests, with the object included in the snapshot code. That'll make it much easier for me to review the precise display decisions.

Reply:
Done! For ones where I reuse the same object a few times for testing different options I just made the snapshot an expression that declared the object as a variable and passed it twice. 
E.g.
```r
  expect_snapshot({
    deep_list <- list(
      list(
        id = "b",
        val = 1,
        children = list(
          list(id = "b1",val = 2.5),
          list(
            id = "b2",
            val = 8,
            children = list(
              list(id = "b21", val = 4)
            )
          )
        )
      ),
      list(id = "a", val = 2)
    )
    tree(deep_list, max_depth = 1)
    tree(deep_list, max_depth = 2)
    tree(deep_list, max_depth = 3)

    tree(deep_list, max_length = 0)
    tree(deep_list, max_length = 2)
    tree(deep_list, max_depth = 1, max_length = 4)
  })
  ```

Works fine but didn't know if that was frowned upon.